phylogeny/prototypeSelection.py

```diff
@@ -448,3 +449,83 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
         prototypes.append(np.asarray(minVals).argmin())
 
     return [dm.ids[idx] for idx in prototypes]
+
+
+def prototype_selection_destructive_maxdist(dm, num_prototypes):
+    '''Heuristically select k prototypes for given distance matrix.
+
+       Prototype selection is NP-hard. This is an implementation of a greedy
+       correctness heuristic: Start with the complete set and interatively
+       remove elements until the number of required prototypes is left.
+       The decision which element shall be removed is based on the minimal
+       distance sum this element has to all other.
+
+    Parameters
+    ----------
+    dm: skbio.stats.distance.DistanceMatrix
+        Pairwise distances for all elements in the full set S.
+    num_prototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+
+    Returns
+    -------
+    list of str
+        A sequence holding selected prototypes, i.e. a sub-set of the
+        elements in the distance matrix.
+
+    Raises
+    ------
+    ValueError
+        The number of prototypes to be found should be at least 2 and at most
+        one element smaller than elements in the distance matrix. Otherwise, a
+        ValueError is raised.
+
+    Notes
+    -----
+    Timing: %timeit -n 100 prototype_selection_constructive_maxdist(dm, 100)
+            100 loops, best of 3: 2.1 s per loop
+            where the dm holds 27,398 elements
+    function signature with type annotation for future use with python >= 3.5:
+    def prototype_selection_constructive_maxdist(dm: DistanceMatrix,
+    num_prototypes: int) -> List[str]:
+    '''
+    if num_prototypes < 2:
+        raise ValueError(("'num_prototypes' must be >= 2, since a single "
+                          "prototype is useless."))
+    if num_prototypes >= len(dm.ids):
+        raise ValueError(("'num_prototypes' must be smaller than the number of"
+                          " elements in the distance matrix, otherwise no "
+                          "reduction is necessary."))
+
+    # clever bookkeeping allows for significant speed-ups!
+
+    # track the number of available elements
+    numRemain = len(dm.ids)
+
+    # distances from each element to all others
+    currDists = dm.data.sum(axis=1)
+
+    # the element to remove first is the one that has smallest distance to all
+    # other. "Removing" works by tagging its distance-sum as infinity. Plus, we
+    # decrease the number of available elements by one.
+    minElmIdx = currDists.argmin()
+    currDists[minElmIdx], numRemain = np.infty, numRemain-1
+
+    # continue until only num_prototype elements are left
+    while (numRemain > num_prototypes):
```

Comment:
I tend to say that you miss something here. It is more than sorting a list and taking the first X elements. In each iteration, some list elements change and thus the ordering is effected: `currDists -= dm.data[minElmIdx]`
I'd be happy to revise the whole loop with a single line of code, but I don't see how.